### PR TITLE
install tap cmd completions

### PR DESCRIPTION
**Pull request by joshka**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [x] Have you successfully run `brew tests` with your changes locally?

-----

Taps can include completion scripts for external commands under
completions/bash, completions/fish, or completions/zsh. brew tap
will automatically install these into the correct directories during
install.

See Homebrew/homebrew-services#73 for a concrete need for completions in a repo.

This PR follows on from https://github.com/Homebrew/brew/pull/1195 but github didn't pick up the extra commits automatically as part of the previous PR (possibly because of my local rebase to pickup changes made later).

### CODE REVIEW

File: Library/Homebrew/cmd/list.rb [hunk 1]
```diff
@@ -76,7 +76,8 @@ def list
     lib/ruby/gems/[12].*
     lib/ruby/site_ruby/[12].*
     lib/ruby/vendor_ruby/[12].*
-    manpages/brew.1
+    manpages/man1/brew.1
+    manpages/man1/brew-cask.1
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/list.rb]: Are there taps using non-`.1` manpages (or wanting to)?

File: Library/Homebrew/cmd/update-report.rb [hunk 2]
```diff
@@ -302,16 +302,11 @@ def migrate_legacy_repository_if_necessary
     $stderr.puts e.backtrace
   end
 
-  def link_completions_and_docs(repository = HOMEBREW_REPOSITORY)
+  def link_additional_files(repository = HOMEBREW_REPOSITORY)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/update-report.rb]: I think perhaps `link_completions_manpages_and_docs` would better signify the intent here.

File: Library/Homebrew/utils.rb [hunk 3]
```diff
@@ -500,8 +500,39 @@ def link_src_dst_dirs(src_dir, dst_dir, command, link_dir: false)
...
+  src_paths = unlink_dir ? [src_dir] : src_dir.find
+  src_paths.each do |src|
+    next if src.directory? && !unlink_dir
+    dst = dst_dir/src.relative_path_from(src_dir)
+    dst.delete if dst.symlink? && src == dst.resolved_path
+    dst.parent.rmdir_if_possible
+  end
+end
+
+def install_manpages(path, command)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/utils.rb]: Can we keep the naming of these to be `(un)link_*`?  Also, probably worth splitting these into a new `utils/link.rb` or similar file.

**Comment by MikeMcQuaid** [comment]
Nice work here so far. A few comments but looking good 👍

File: Library/Homebrew/cmd/list.rb (hunk 1 above)
Comment (joshka) [inline Library/Homebrew/cmd/list.rb]: This allows us to handle the man pages just as a generic nested folder for install / uninstall rather than specifically handling the man case. I'd like to leave this like this for simplicity. It makes it obvious where a non.1 case would belong (arguablely brew should be .8)

File: Library/Homebrew/cmd/update-report.rb (hunk 2 above)
Comment (joshka) [inline Library/Homebrew/cmd/update-report.rb]: But what happens when we add a config file, some other related thing, ...

File: Library/Homebrew/utils.rb (hunk 3 above)
Comment (joshka) [inline Library/Homebrew/utils.rb]: I like install because it tells the intent rather than the effect. How this occurs might change without too much hassle, but why it's happening remains the same

Comment (joshka) [inline Library/Homebrew/utils.rb]: Kinda agree on link.rb

File: Library/Homebrew/cmd/list.rb (hunk 1 above)
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/list.rb]: Do we have any non-`.1` cases currently, though? If not, I'd rather leave as-is (or at least make that another PR).

File: Library/Homebrew/cmd/update-report.rb (hunk 2 above)
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/update-report.rb]: 1) YAGNI, 2) we can change the name then. `link_additional_files` doesn't really communicate anything.

File: Library/Homebrew/utils.rb (hunk 3 above)
Comment (MikeMcQuaid) [inline Library/Homebrew/utils.rb]: Let's leave it as `link` but move these to `link.rb` 👍

File: Library/Homebrew/cmd/list.rb (hunk 1 above)
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/list.rb]: @joshka Ping on this, let's leave it as-is for now, please and save moving these around to another PR.

File: Library/Homebrew/cmd/update-report.rb (hunk 2 above)
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/update-report.rb]: @joshka Let's use `link_completions_manpages_and_docs`, thanks.

File: Library/Homebrew/utils.rb (hunk 3 above)
Comment (MikeMcQuaid) [inline Library/Homebrew/utils.rb]: Again, let's leave these as `link`, thanks.

**Comment by MikeMcQuaid** [comment]
@joshka Ping on the changes/conflicts here, thanks.

**Comment by joshka** [comment]
@MikeMcQuaid I had a rather extended Christmas / new year vacation and then moved house and had to put this on the back burner. I'll take a look later today and get it back on track. :)

**Comment by joshka** [comment]
@MikeMcQuaid  I've fixed up the review comments. I'm happy to rebase / squash / resolve conflicts once you've looked at it.

**Comment by joshka** [comment]
I've rebased against master, squashed the WIP commits, and fixed the review item that I missed (the link functions are now in link.rb)

File: Library/Homebrew/cmd/update-report.rb [hunk 4]
```diff
@@ -302,16 +303,11 @@ def migrate_legacy_repository_if_necessary
...
-                      HOMEBREW_PREFIX/"etc/bash_completion.d", command)
-    link_src_dst_dirs(repository/"docs",
-                      HOMEBREW_PREFIX/"share/doc/homebrew", command, link_dir: true)
-    link_src_dst_dirs(repository/"completions/zsh",
-                      HOMEBREW_PREFIX/"share/zsh/site-functions", command)
-    link_src_dst_dirs(repository/"manpages",
-                      HOMEBREW_PREFIX/"share/man/man1", command)
+    link_completions(repository, command)
+    link_manpages(repository, command)
+    link_docs(repository, command)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/update-report.rb]: Given this function is only called from here and only applies to Homebrew/brew I wonder if the indirection here is a little confusing (but have no strong views either way).

File: Library/Homebrew/link.rb [hunk 5]
```diff
@@ -0,0 +1,62 @@
+def link_src_dst_dirs(src_dir, dst_dir, command, link_dir: false)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/link.rb]: This should be in `utils/link.rb` and be inside a `module Link`

File: Library/Homebrew/link.rb [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+  src_paths = unlink_dir ? [src_dir] : src_dir.find
+  src_paths.each do |src|
+    next if src.directory? && !unlink_dir
+    dst = dst_dir/src.relative_path_from(src_dir)
+    dst.delete if dst.symlink? && src == dst.resolved_path
+    dst.parent.rmdir_if_possible
+  end
+end
+
+def link_manpages(path, command)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/link.rb]: I realise I'm changing my mind but could always be `link_man1pages` or something similar if you wanted to communicate this only refers to `man1`

File: Library/Homebrew/link.rb (hunk 5 above)
Comment (MikeMcQuaid) [inline Library/Homebrew/link.rb]: Additionally, this method could be `private.

File: Library/Homebrew/link.rb [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+  return if conflicts.empty?
+  onoe <<-EOS.undent
+    Could not link:
+    #{conflicts.join("\n")}
+
+    Please delete these paths and run `#{command}`.
+  EOS
+end
+
+def unlink_src_dst_dirs(src_dir, dst_dir, unlink_dir: false)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/link.rb]: This method could be `private`

File: Library/Homebrew/cmd/update-report.rb (hunk 4 above)
Comment (joshka) [inline Library/Homebrew/cmd/update-report.rb]: I'm not sure that I understand your question. Are you talking about `link_completions_manpages_and_docs()` or link_docs() method here?
If it's the former, the method exists to DRY up the same code in the taps.
If it's the latter, the method exists for consistency with the other `link_` methods.

Or perhaps you're suggesting to inline the calls to `link_*` at the call sites? I'm a fan of smaller methods rather than larger generally, and we need to call `Tap.link_completions_and_manpages` externally so I think for consistency this means this stays.

Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/update-report.rb]: `link_docs`, yes. If you feel strongly: I don't so leave as-is.

File: Library/Homebrew/link.rb (hunk 6 above)
Comment (joshka) [inline Library/Homebrew/link.rb]: I think it makes enough sense as-is to leave it and only update the implementation if a situation presents that other man sections are needed. Adding manX would likely mean reorganizing a bunch of folders across various repos, so it would be a fairly heavy change that isn't just a matter of adding a new method

**Comment by joshka** [comment]
Addressed module change

File: Library/Homebrew/utils/link.rb [hunk 8]
```diff
@@ -0,0 +1,68 @@
+module Link
```
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/link.rb]:
```ruby
module Utils
  module Link
```

**Comment by MikeMcQuaid** [comment]
Final bit of pedantry but otherwise looking good. Once that's addressed I'll give this some testing and get Homebrew/homebrew-services ready to go with it then 🚢.

File: Library/Homebrew/utils/link.rb (hunk 8 above)
Comment (joshka) [inline Library/Homebrew/utils/link.rb]: Cool with me. I wasn't sure about the idiomatic ruby way on this was and the modules under `/utils` seemed to follow both approaches.

**Comment by joshka** [comment]
Done. I note that while this PR adds more tests as an absolute number, the code coverage check on the diff reports that it's not quite enough. I think this is due to a large portion of this diff being a refactor of methods to a new module.

**Comment by MikeMcQuaid** [comment]
Thanks for your work and patience here @joshka!
